README.md
# Skeleton 骨架屏
基于 Taro 的一个简单易用的骨架屏组件

> **2.0**版本以上支持 **Taro3.0**
### 支持多端平台使用 
1. 微信小程序
2. h5
3. 百度小程序
4. 头条小程序
5. 支付宝小程序
6. 其它平台未测试

### 快速上手
#### 引入组件

``` javascript
import Skeleton from 'taro-skeleton'
```
#### 引入组件样式的方式
* 全局引入（JS中）
``` js
import 'taro-skeleton/dist/index.css' // 引入组件样式
```

### 配置需要额外编译的源码模块
由于引用 `node_modules` 的模块，默认不会编译，所以需要额外给 H5 配置 `esnextModules`，在 taro 项目的 `config/index.js` 中新增如下配置项：
``` js
h5: {
  esnextModules: ['taro-skeleton']
}
```

## 代码演示

### 基础用法

通过`title`属性显示标题占位图，通过`row`属性配置占位段落行数

``` html
<Skeleton title row={3} />
```

### 显示头像

通过`avatar`属性显示头像占位图

``` html
<Skeleton title avatar row={3} />
```
### 不同排列方式
通过`type`属性来控制排列方式，默认值为 `row`，可选`column`
``` html
<Skeleton type='column' title titleWidth={'80%'} avatar />
```
### 自定义样式
通过添加自定义`className`类名进行覆盖即可`
``` scss
// index.scss
.custom-class { 
  .skeleton-avatar, .skeleton-row, .skeleton-title {
    background-color: red;
  }
}
```
``` jsx
// index.jsx
<Skeleton className='custom-class' row={1} rowProps={[{
  width: '50%',
  height: '50px'
}]} avatar
></Skeleton>
```

### 展示子组件

将`loading`属性设置成`false`表示内容加载完成，此时会隐藏占位图，并显示`Skeleton`的子组件

``` html
<Skeleton
  title
  avatar
  row={3}
  loading={loading}
>
  <Text>实际内容</Text>
</Skeleton>
```

```jsx
export default class Index extends Component {
  state = {
    loading: false
  }
  render () {
    return (
      <View className='index'>
        <Skeleton loading={this.state.loading} title avatar row={2} action></Skeleton>
      </View>
    )
  }
}
```

## 效果截图
![image.gif](https://img10.360buyimg.com/img/jfs/t1/60232/20/13975/296420/5db7e244E05a0d555/d7724d0dd3af11ec.gif)



## API

### Props

| 参数 | 说明 | 类型 | 默认值 | 版本 |
|------|------|------|------|------|
| type | 定义排列方式  | `row/column` | `row` | 1.0.12 |
| row | 段落占位图行数 | `number` | `0` | - |
| rowWidth | 段落占位图宽度，可传数组来设置每一行的宽度 | `number/string/number[]/string[]` | `100%` | - |
| rowHeight | 段落占位图高度，可传数组来设置每一行的高度 | `number/string/number[]/string[]` | `24` | 1.0.7 |
| rowProps | 用于定制 row 的宽跟高，可传数组来设置每一行的宽跟高，如果配置了该属性，则 row-height 配置无效 | `RowProps/RowProps[]` | - | 1.0.7 |
| title | 是否显示标题占位图 | `boolean` | `false` | - |
| titleWidth | 标题占位图宽度 | `number/string` | `40%` | - |
| avatar | 是否显示头像占位图 | `boolean` | `false` | - |
| avatarSize | 头像占位图大小 | `number/string` | `90` | - |
| avatarShape | 头像占位图形状，可选值为`square` | `string` | `round` | - |
| action | 显示右边操作按钮占位图 | `boolean` | `false` | - |
| loading | 是否显示占位图，传`false`时会展示子组件内容 | `boolean` | `true` | - 
| animate | 是否开启动画 | `boolean` | `true` | - |
| animateName | 动画类型，可选值为`elastic` | `string` | `blink` | 1.3.9 |
| contentAlignStyle | 内部内容对齐方式，可选值为 | `left/center/right` | `center` | 1.4.0
| designWidth |Taro.pxTransform(px, designWidth) 的designWidth的属性| `number` | `750` | 2.0.0 |
### RowProps

| 参数 | 说明 | 类型 | 默认值 | 版本 |
|------|------|------|------|------|
| width | 段落占位图宽数 | `number/string` | - | - |
| height | 段落占位图高度 | `number/string` | - | - |

## [更新日志](https://github.com/lentoo/taro-skeleton/blob/master/CHANGELOG.md)
## ? 


CHANGELOG.md
## [2.0.3](https://github.com/lentoo/taro-skeleton/compare/2.0.2...2.0.3) (2020-10-22)


### Bug Fixes

* 🐛 类型声明 ([83a30a9](https://github.com/lentoo/taro-skeleton/commit/83a30a919be6ce253cb81a2106ae707db221cae3))

# [2.0.0](https://github.com/lentoo/taro-skeleton/compare/1.2.2...2.0.0) (2020-09-19)


### Features

* 🎸 支持Taro3.0版本 ([7705105](https://github.com/lentoo/taro-skeleton/commit/77051052ab89f3d759a374a1286a0ffc4757a037)), closes [#20](https://github.com/lentoo/taro-skeleton/issues/20) [#21](https://github.com/lentoo/taro-skeleton/issues/21)
* 🎸 组件优化 ([a3337a1](https://github.com/lentoo/taro-skeleton/commit/a3337a11b474a3cd58d5135e36f0e54c8eeb2f8e))



## [1.2.2](https://github.com/lentoo/taro-skeleton/compare/1.2.1...1.2.2) (2020-08-08)



## [1.2.1](https://github.com/lentoo/taro-skeleton/compare/1.2.0...1.2.1) (2020-08-08)



# [1.2.0](https://github.com/lentoo/taro-skeleton/compare/1.1.4...1.2.0) (2020-06-04)


### Features

* "新增自定义类名和部分样式调整" ([5826ef2](https://github.com/lentoo/taro-skeleton/commit/5826ef2ee0d82622c56bf5440e1eac942b3ec39b))
* 🎸 优化自定义类名，补充文档 ([ffa6e31](https://github.com/lentoo/taro-skeleton/commit/ffa6e31248d432e24d1b6c14fec40340f95f4a59)), closes [#15](https://github.com/lentoo/taro-skeleton/issues/15)



## [1.1.4](https://github.com/lentoo/taro-skeleton/compare/1.1.3...1.1.4) (2020-04-08)


### Bug Fixes

* 🐛 删除多渲染的一个View标签。改为空标签 ([9607939](https://github.com/lentoo/taro-skeleton/commit/9607939470b1f2947f3730bc5830b6bd6d12fd85))



## [1.1.3](https://github.com/lentoo/taro-skeleton/compare/1.1.0...1.1.3) (2020-04-07)



# [1.1.0](https://github.com/lentoo/taro-skeleton/compare/1.0.13...1.1.0) (2020-03-30)



## [1.0.13](https://github.com/lentoo/taro-skeleton/compare/1.0.12...1.0.13) (2019-12-25)



## [1.0.12](https://github.com/lentoo/taro-skeleton/compare/1.0.11-0...1.0.12) (2019-10-29)


### Features

* 🎸 新增纵向排列方式 ([9b75e54](https://github.com/lentoo/taro-skeleton/commit/9b75e54d1d83903cec6dccb7f5b149524e3d6673)), closes [#3](https://github.com/lentoo/taro-skeleton/issues/3) [#5](https://github.com/lentoo/taro-skeleton/issues/5)



## [1.0.11-0](https://github.com/lentoo/taro-skeleton/compare/1.0.10...1.0.11-0) (2019-09-02)


### Features

* 🎸 新增 pre-commit 功能 ([0d4b66e](https://github.com/lentoo/taro-skeleton/commit/0d4b66e22f5007c6a7a45e58fc3df9d4e9208ec8))



## [1.0.10](https://github.com/lentoo/taro-skeleton/compare/1.0.9...1.0.10) (2019-09-02)


### Bug Fixes

* 🐛 修复提示npm包不存在taro-skeleton ([e9e03c7](https://github.com/lentoo/taro-skeleton/commit/e9e03c74764117e1fc03bdec5093724e7e654e27)), closes [#2](https://github.com/lentoo/taro-skeleton/issues/2)



## [1.0.9](https://github.com/lentoo/taro-skeleton/compare/1.0.8...1.0.9) (2019-08-28)


### Features

* 🎸 新增骨架屏组件动画类型 ([e5b864b](https://github.com/lentoo/taro-skeleton/commit/e5b864b2fd05d9e69b5f914c320cf942bef73008))



## [1.0.8](https://github.com/lentoo/taro-skeleton/compare/1.0.7...1.0.8) (2019-08-16)



## [1.0.7](https://github.com/lentoo/taro-skeleton/compare/1.0.6...1.0.7) (2019-08-16)


### Features

* 🎸 支持 row 占位图自定义宽跟高 ([d3003f5](https://github.com/lentoo/taro-skeleton/commit/d3003f5197633da973296be77f0dbc60bed6ac3e)), closes [#1](https://github.com/lentoo/taro-skeleton/issues/1)



## [1.0.6](https://github.com/lentoo/taro-skeleton/compare/1.0.5...1.0.6) (2019-08-14)



## [1.0.5](https://github.com/lentoo/taro-skeleton/compare/1.0.4...1.0.5) (2019-08-14)


### Bug Fixes

* 🐛 row add key ([2cee2c4](https://github.com/lentoo/taro-skeleton/commit/2cee2c40b6480956769ffcbec3714376416198ab))



## [1.0.4](https://github.com/lentoo/taro-skeleton/compare/1.0.3...1.0.4) (2019-08-14)


### Bug Fixes

* 🐛 修改 SkeletonProps 参数 ([21a9147](https://github.com/lentoo/taro-skeleton/commit/21a91471df6b2a63d7e0067ca81694ea38a96399))



## [1.0.3](https://github.com/lentoo/taro-skeleton/compare/7d6d48c9f2ce858efc5ae79f5a2d0ac683bdafee...1.0.3) (2019-08-14)


### Features

* 🎸 新增骨架屏组件 ([3ee7849](https://github.com/lentoo/taro-skeleton/commit/3ee7849111e0e95a6f906aa37572f30f9d471596))
* 🎸 init project ([7d6d48c](https://github.com/lentoo/taro-skeleton/commit/7d6d48c9f2ce858efc5ae79f5a2d0ac683bdafee))





config/index.js
// eslint-disable-next-line import/no-commonjs
const path = require('path')

const config = {
  projectName: 'taro-skeleton',
  date: '2019-8-14',
  designWidth: 750,
  deviceRatio: {
    '640': 2.34 / 2,
    '750': 1,
    '828': 1.81 / 2
  },
  sourceRoot: 'src',
  outputRoot: 'dist',
  babel: {
    sourceMap: true,
    presets: [
      ['env', {
        modules: false
      }]
    ],
    plugins: [
      'transform-decorators-legacy',
      'transform-class-properties',
      'transform-object-rest-spread'
    ]
  },
  defineConstants: {
  },
  copy: {
    patterns: [
    ],
    options: {
    }
  },
  mini: {
    webpackChain () {},
    cssLoaderOption: {},
    postcss: {
      pxtransform: {
        enable: true,
        config: {}
      },
      url: {
        enable: true,
        config: {
          limit: 10240 // 设定转换尺寸上限
        }
      }
    }
  },
  h5: {
    publicPath: '/',
    staticDirectory: 'static',
    webpackChain () {},
    postcss: {
      autoprefixer: {
        enable: true,
        config: {
          browsers: [
            'last 3 versions',
            'Android >= 4.1',
            'ios >= 8'
          ]
        }
      },
      pxtransform: {
        enable: true,
        config: {}
      }
    }
  }
}
if (process.env.TARO_BUILD_TYPE === 'ui') {
  Object.assign(config.h5, {
    enableSourceMap: false,
    enableExtract: false,
    enableDll: false
  })
  config.h5.webpackChain = chain => {
    chain.plugins.delete('htmlWebpackPlugin')
    chain.plugins.delete('addAssetHtmlWebpackPlugin')
    chain.merge({
      output: {
        path: path.join(process.cwd(), 'dist', 'h5'),
        filename: 'index.js',
        libraryTarget: 'umd',
        library: 'taro-skeleton'
      },
      externals: {
        nervjs: 'commonjs2 nervjs',
        classnames: 'commonjs2 classnames',
        '@tarojs/components': 'commonjs2 @tarojs/components',
        '@tarojs/taro-h5': 'commonjs2 @tarojs/taro-h5',
        'weui': 'commonjs2 weui'
      }
    })
  }
}
module.exports = function (merge) {
  if (process.env.NODE_ENV === 'development') {
    return merge({}, config, require('./dev'))
  }
  return merge({}, config, require('./prod'))
}


config/rollup.config.js
import NodePath from 'path'
import RollupJson from '@rollup/plugin-json'
import RollupNodeResolve from '@rollup/plugin-node-resolve'
import RollupCommonjs from '@rollup/plugin-commonjs'
import RollupTypescript from 'rollup-plugin-typescript2'
import RollupCopy from 'rollup-plugin-copy'
import RollupScss from 'rollup-plugin-scss'
import Package from '../package.json'
const externalPackages = [
  'react',
  'react-dom',
  '@tarojs/components',
  '@tarojs/runtime',
  '@tarojs/taro',
  '@tarojs/react'
]
const resolveFile = path => NodePath.resolve(__dirname, '..', path)

export default {
  input: resolveFile(Package.source),
  output: [
    {
      file: resolveFile(Package.main),
      format: 'cjs',
      sourcemap: true
    },
    {
      file: resolveFile(Package.module),
      format: 'es',
      sourcemap: true
    },
    {
      file: resolveFile(Package.browser),
      format: 'umd',
      name: 'taro-skeleton',
      sourcemap: true,
      globals: {
        react: 'React',
        '@tarojs/components': 'components',
        '@tarojs/taro': 'Taro'
      }
    }
  ],
  external: externalPackages,
  plugins: [
    RollupScss(),
    RollupNodeResolve({
      customResolveOptions: {
        moduleDirectory: 'node_modules'
      }
    }),
    RollupCommonjs({
      include: /\/node_modules\//
    }),
    RollupJson(),
    RollupTypescript({
      tsconfig: resolveFile('tsconfig.rollup.json')
    }),
    RollupCopy({
      targets: [
        {
          src: resolveFile('src/style'),
          dest: resolveFile('dist')
        }
      ]
    })
  ]
}

types/index.d.ts
import { FC, ReactChildren } from 'react'
/**
 * @description Row 属性的宽高
 * @author lentoo
 * @date 2019-08-16
 * @export
 * @interface RowProps
 */
export interface RowProps {
  width: string | number
  height: string | number
}
export type AnimateName = 'blink' | 'elastic'
export type AvatarShapeOptions = 'round' | 'square'

/**
 * @description 骨架屏组件参数
 * @author lentoo
 * @date 2019-08-16
 * @export
 * @interface SkeletonProps
 */
export interface SkeletonProps {
  className?: string
  /**
   * @description 排列方向  横向 或者 纵向， 默认 row
   * @type {('row' | 'column')}
   * @memberof SkeletonProps
   */
  type?: 'row' | 'column'
  /**
   * @description 段落占位图行数
   * @type {number}
   */
  row?: number
  /**
   * @description 是否显示占位图，传 `false` 时会展示子组件内容
   * @type {boolean}
   */
  loading?: boolean
  /**
   * @description 是否显示标题占位图
   * @type {boolean}
   */
  title?: boolean
  /**
   * @description 标题占位图宽度
   * @type {(string | number)}
   */
  titleWidth?: string | number
  /**
   * @description 是否显示头像占位图
   * @type {boolean}
   */
  avatar?: boolean
  /**
   * @description avatar-size
   * @type {number}
   */
  avatarSize?: number
  /**
   * @description 头像占位图形状，可选值为 `square` 、`round` 默认值：round
   * @type {AvatarShapeOptions}
   */
  avatarShape?: AvatarShapeOptions
  /**
   * @description 是否显示右边操作按钮占位图
   * @type {boolean}
   */
  action?: boolean
  /**
   * @description 是否开启动画
   * @type {boolean}
   */
  animate?: boolean
  /**
   * @description 动画名称
   * @type {AnimateName}
   * @memberof SkeletonProps
   */
  animateName?: AnimateName
  /**
   * @description 段落占位图宽度，可传数组来设置每一行的宽度
   * @type {(number | string | (number | string)[])}
   */
  rowWidth?: number | string | (number | string)[]
  /**
   * @description 段落占位图高度，可传数组来设置每一行的高度
   * @type {(number | string | (number | string)[])}
   * @memberof SkeletonProps
   */
  rowHeight?: number | string | (number | string)[]
  /**
   * @description 用于定制 row 的宽跟高，可传数组来设置每一行的宽跟高，如果配置了该属性，则 rowWidth 配置无效
   * @type {(RowProps | RowProps[])}
   * @memberof SkeletonProps
   */
  rowProps?: RowProps | RowProps[]
  /**
   * @description 子组件内容
   * @type {JSX.Element}
   */
  /**
   * @description skeleton-content的对齐方式，默认center
   * @type {('left' | 'center' | 'right')}
   */
  contentAlignStyle?: 'left' | 'center' | 'right'

  /**
   * @description 自定义类名
   * @deprecated 废弃
   * @type String
   * @link http://taro-docs.jd.com/taro/docs/component-style
   */
  'skeleton-custom-class'?: string

  designWidth: number

  children?: JSX.Element
}

declare const Skeleton: FC<SkeletonProps>

export default Skeleton